Add multi-row insert support to SQL builder

sivdead opened the pull request:
make `AbstratSQL.INSERT_VALUES` could be invoke multiple times to  do batch insert.

harawata commented:
Thank you for the PR, @sivdead !

It could break backward compatibility, though.
Please see the following test case. It will fail if this PR is merged.

```java
  @Test
  public void variableLengthArgumentOnIntoColumnsAndValues() {
    final String sql = new SQL() {{
      INSERT_INTO("TABLE_A").INTO_COLUMNS("a", "b").INTO_VALUES("#{a}").INTO_VALUES("#{b}");
    }}.toString();
    assertEquals("INSERT INTO TABLE_A\n (a, b)\nVALUES (#{a}, #{b})", sql);
  }
```
I'm not sure what's the best approach here, but it might be better to add a new method instead of modifying `INTO_VALUES`.

Cc-ing @h3adache @kazuki43zoo

kazuki43zoo commented:
>  it might be better to add a new method instead of modifying INTO_VALUES.

I agree with it.

For example, one approach is as follow:

```java
  @Test
  public void multipleRowsInsert() {
    final String sql = new SQL() {{
      INSERT_INTO("TABLE_A")
          .INTO_COLUMNS("a", "b")
          .INTO_VALUES("#{a1}", "#{b1}")
          .ADD_ROW()
          .INTO_VALUES("#{a2}", "#{b2}")
      ;
    }}.toString();

    assertEquals("INSERT INTO TABLE_A\n (a, b)\nVALUES \n (#{a1}, #{b1}),\n (#{a2}, #{b2})", sql);
  }
```

NOTE: Multiple rows insert style is different by database vender and versions.

sivdead commented:
> > it might be better to add a new method instead of modifying INTO_VALUES.
> 
> I agree with it.
> 
> For example, one approach is as follow:
> 
> ```java
>   @Test
>   public void multipleRowsInsert() {
>     final String sql = new SQL() {{
>       INSERT_INTO("TABLE_A")
>           .INTO_COLUMNS("a", "b")
>           .INTO_VALUES("#{a1}", "#{b1}")
>           .ADD_ROW()
>           .INTO_VALUES("#{a2}", "#{b2}")
>       ;
>     }}.toString();
> 
>     assertEquals("INSERT INTO TABLE_A\n (a, b)\nVALUES \n (#{a1}, #{b1}),\n (#{a2}, #{b2})", sql);
>   }
> ```
> 
> NOTE: Multiple rows insert style is different by database vender and versions.

thanks for the advice,so I add a method ADD_ROW to avoid that multi INTO_VALUES in single insert problem.
For example:
```java
  @Test
  public void batchInsertWithMultipleInsertValues(){
    final String sql = new SQL() {{
      INSERT_INTO("TABLE_A");
      INTO_COLUMNS("a", "b");
      INTO_VALUES("#{a1}");
      INTO_VALUES("#{b1}");
      ADD_ROW();
      INTO_VALUES("#{a2}");
      INTO_VALUES("#{b2}");
    }}.toString();
    System.out.println(sql);
    assertEquals("INSERT INTO TABLE_A\n (a, b)\nVALUES \n (#{a1}, #{b1}),\n (#{a2}, #{b2})", sql);
  }
```
I see that batch insert in oracle is different than others,but I don't know how to get database type or version in sqlBuilder-,-

harawata commented:
Thank you for the update, @sivdead !

@kazuki43zoo ,
Does `ADD_ROW()` look OK to you?
According to [Wikipedia](https://en.wikipedia.org/wiki/Insert_(SQL)#Multirow_inserts), this syntax is the standard one, so I personally think it's OK even if it does not work with all DBs.

I may request a few minor code changes later.

kazuki43zoo commented:
Hi @harawata , sorry for late reply.

> Does ADD_ROW() look OK to you?

I think it is OK.

> According to Wikipedia, this syntax is the standard one, so I personally think it's OK even if it does not work with all DBs.

I thinks so, too!

sivdead commented:
> Hi @sivdead ,
> I'm sorry about the delay! There were some high priority issues.
> I have added a few comments.
> Please take a look when you have time.
> 
> @kazuki43zoo ,
> Thanks! Feel free to add comments if you noticed anything.



> Hi @sivdead ,
> I'm sorry about the delay! There were some high priority issues.
> I have added a few comments.
> Please take a look when you have time.
> 
> @kazuki43zoo ,
> Thanks! Feel free to add comments if you noticed anything.

thanks for correcting my mistake! and also sorry about the delay reply :(

sivdead commented:
> > Hi @sivdead ,
> > I'm sorry about the delay! There were some high priority issues.
> > I have added a few comments.
> > Please take a look when you have time.
> > @kazuki43zoo ,
> > Thanks! Feel free to add comments if you noticed anything.
> 
> > Hi @sivdead ,
> > I'm sorry about the delay! There were some high priority issues.
> > I have added a few comments.
> > Please take a look when you have time.
> > @kazuki43zoo ,
> > Thanks! Feel free to add comments if you noticed anything.
> 
> thanks for correcting my mistake! and also sorry about the delay reply :(

I'm not sure if i should recommit with the code you suggested or let u do this :(

harawata commented:
Thank you for the update, @sivdead !
I have resolved the conflict and made a few minor adjustment.

@kazuki43zoo , 
Could you do the final check and merge if everything is OK?
Thanks in advance!

kazuki43zoo commented on src/main/java/org/apache/ibatis/jdbc/AbstractSQL.java:
```diff
@@ -310,6 +314,11 @@ public T OFFSET(long value) {
     return getSelf();
   }
 
+  public T ADD_ROW() {
```
Please add `@Since 3.5.2`

kazuki43zoo replied:
- [ ] Please check a position of JavaDoc for `ADD_ROW` (Probably, there is mistake for your merging).

- [ ] JavaDoc's item order is not standard. Standard order is "description -> `@since`" in this case. But currently "`@since` -> description".

- [ ] Please add `"."`(period) end with sentence of description.

- [ ] Please change to `@since` from `@Since` (`S` change to lower case)

kazuki43zoo commented on src/main/java/org/apache/ibatis/jdbc/AbstractSQL.java:
```diff
@@ -374,13 +383,14 @@ public boolean isEmpty() {
     List<String> lastList = new ArrayList<>();
     List<String> columns = new ArrayList<>();
-    List<String> values = new ArrayList<>();
+    List<List<String>> values = new ArrayList<>();
```
I prefer the `valuesList` rather than `values` in this case.

kazuki43zoo replied:
I suggest to change `valuesList` rather than `valueList` because `valueList` same mean with `values`.

sivdead replied:
have it fixed,so sad for that merge mistake.

kazuki43zoo commented on src/main/java/org/apache/ibatis/jdbc/AbstractSQL.java:
```diff
@@ -78,7 +78,11 @@ public T INTO_COLUMNS(String... columns) {
+    List<List<String>> list = sql().values;
+    int index = list.size() - 1;
+    for (String value : values) {
+      list.get(index).add(value);
```
`list.get(index)` can be move outside the `for` phrase.

sivdead commented:
> @sivdead
> 
> I've added some small comments. If you have a time, please check it.

i have do as you suggested,again thank you for helping me a lot on this pr :)

kazuki43zoo commented on src/main/java/org/apache/ibatis/jdbc/AbstractSQL.java:
```diff
@@ -360,27 +373,28 @@ public boolean isEmpty() {
     }
 
     StatementType statementType;
-    List<String> sets = new ArrayList<>();
```
Changes of 3502f7ced8ec30f6196c4253df4d22acba601c40 (merging from latest master branch by @harawata) was reverted. Probably your merging operation is incorrect. In this case, the code that uses the diamond operator is latest and correct implementation.

kazuki43zoo replied:
@sivdead This comment does not fixed. Why?

sivdead replied:
@kazuki43zoo please recheck it ,maybe i used an wrong branch

kazuki43zoo replied:
@sivdead I've fixed it and merged at now! Thanks for your contributing!

sivdead commented:
> @sivdead Thanks for your rework! But I have some change requests. Please check new comment and replying comments.

get it fixed
